File: jscomp/test/bs_stack_test.js
'use strict';

var Js_null = require("../../lib/js/js_null.js");
var Bs_Queue = require("../../lib/js/bs_Queue.js");
var Bs_Stack = require("../../lib/js/bs_Stack.js");

function inOrder(v) {
  var current = v;
  var s = {
    root: null
  };
  var q = Bs_Queue.create(/* () */0);
  while(current !== null) {
    var v$1 = current;
    Bs_Stack.push(s, v$1);
    current = v$1.left;
  };
  while(s.root !== null) {
    current = Bs_Stack.popNull(s);
    var v$2 = current;
    Bs_Queue.addDone(q, v$2.value);
    current = v$2.right;
    while(current !== null) {
      var v$3 = current;
      Bs_Stack.push(s, v$3);
      current = v$3.left;
    };
  };
  return Bs_Queue.toArray(q);
}

function inOrder3(v) {
  var current = v;
  var s = {
    root: null
  };
  var q = Bs_Queue.create(/* () */0);
  while(current !== null) {
    var v$1 = current;
    Bs_Stack.push(s, v$1);
    current = v$1.left;
  };
  Bs_Stack.dynamicPopIter(s, (function (popped) {
          Bs_Queue.addDone(q, popped.value);
          var current = popped.right;
          while(current !== null) {
            var v = current;
            Bs_Stack.push(s, v);
            current = v.left;
          };
          return /* () */0;
        }));
  return Bs_Queue.toArray(q);
}

function inOrder2(v) {
  var todo = /* true */1;
  var cursor = v;
  var s = {
    root: null
  };
  var q = Bs_Queue.create(/* () */0);
  while(todo) {
    if (cursor !== null) {
      var v$1 = cursor;
      Bs_Stack.push(s, v$1);
      cursor = v$1.left;
    } else if (s.root !== null) {
      cursor = Bs_Stack.popNull(s);
      var current = cursor;
      Bs_Queue.addDone(q, current.value);
      cursor = current.right;
    } else {
      todo = /* false */0;
    }
  };
  return /* () */0;
}

function n(l, r, a) {
  return {
          value: a,
          left: Js_null.fromOption(l),
          right: Js_null.fromOption(r)
        };
}

var test1 = n(/* Some */[n(/* Some */[n(/* None */0, /* None */0, 4)], /* Some */[n(/* None */0, /* None */0, 5)], 2)], /* Some */[n(/* None */0, /* None */0, 3)], 1);

function pushAllLeft(st1, s1) {
  var current = st1;
  while(current !== null) {
    var v = current;
    Bs_Stack.push(s1, v);
    current = v.left;
  };
  return /* () */0;
}

var test2 = n(/* Some */[n(/* Some */[n(/* Some */[n(/* Some */[n(/* None */0, /* None */0, 4)], /* None */0, 2)], /* None */0, 5)], /* None */0, 1)], /* None */0, 3);

var test3 = n(/* Some */[n(/* Some */[n(/* Some */[n(/* None */0, /* None */0, 4)], /* None */0, 2)], /* None */0, 5)], /* Some */[n(/* None */0, /* None */0, 3)], 1);

console.log(inOrder(test1));

console.log(inOrder3(test1));

var S = 0;

var Q = 0;

exports.S = S;
exports.Q = Q;
exports.inOrder = inOrder;
exports.inOrder3 = inOrder3;
exports.inOrder2 = inOrder2;
exports.n = n;
exports.test1 = test1;
exports.pushAllLeft = pushAllLeft;
exports.test2 = test2;
exports.test3 = test3;
/* test1 Not a pure module */

File: jscomp/test/bs_stack_test.ml


type node = {
  value : int ; 
  left : t;
  right : t
} 
and t = node Js.undefined
[@@bs.deriving abstract]



module S = Belt.MutableStack
module Q = Belt.MutableQueue 

let inOrder (v : t) = 
  let current =  ref v in 
  let s : node S.t = S.make () in 
  let q : int Q.t = Q.make () in 
  while !current != Js.undefined do 
    let v = Js.Undefined.getUnsafe !current  in 
    S.push s v; 
    current := leftGet v; 
  done ;
  while not (S.isEmpty s ) do 
    current := S.popUndefined s ;
    let v = Js.Undefined.getUnsafe !current in 
    Q.add q (valueGet v);
    current := rightGet v ;
    while !current != Js.undefined do 
      let v = Js.Undefined.getUnsafe !current  in 
      S.push s v; 
      current := leftGet v; 
    done ;    
  done; 
  Q.toArray q 

let inOrder3 (v : t) = 
  let current =  ref v in 
  let s : node S.t = S.make () in 
  let q : int Q.t = Q.make () in 
  while !current != Js.undefined do 
    let v = Js.Undefined.getUnsafe !current  in 
    S.push s v; 
    current := leftGet v; 
  done ;
  S.dynamicPopIter s begin fun  popped -> 
    Q.add q (valueGet popped);
    let current = ref (rightGet popped) in 
    while !current != Js.undefined do 
      let v = Js.Undefined.getUnsafe !current in 
      S.push s v;  
      current := leftGet v
    done 
  end;
  Q.toArray q 

let inOrder2 (v : t) =   
  let todo = ref true in 
  let cursor = ref v in 
  let s : node S.t = S.make () in 
  let q : int Q.t = Q.make () in 
  while !todo do 
    if !cursor != Js.undefined then 
      (
        let v = (Js.Undefined.getUnsafe !cursor) in 
        S.push s v;
        cursor := leftGet v)
    else 
      begin 
        if not (S.isEmpty s) then 
          (cursor := S.popUndefined s ;
           let current = Js.Undefined.getUnsafe !cursor in 
           Q.add q (valueGet current);
           cursor := rightGet current)
        else 
          todo := false
      end 
  done 

let n
    ?l ?r a =
  node   ~value:a 
    ~left:(Js.Undefined.fromOption l) 
    ~right:(Js.Undefined.fromOption r)


let test1 = 
  n 1
    ~l:
      (n 2 
         ~l:(n 4 )
         ~r:(n 5))
    ~r:(n 3)

let pushAllLeft st1 s1 = 
  let current = ref st1 in 
  while !current != Js.undefined do 
    let v = Js.Undefined.getUnsafe !current  in 
    S.push s1 v; 
    current := leftGet v; 
  done 
;;


let test2 = 
  n 3
  ~l:(
    n 1 
    ~l:(
      n 5
      ~l:
      (n 2
      ~l:
      (n 4)
      )
    )
  )

let test3 = 
  
    n 1 
    ~l:(
      n 5
      ~l:
      (n 2
      ~l:
      (n 4)
      )
    )
  ~r:(n 3)

let () =     
  Js.log (inOrder (Js.Undefined.return test1));
  Js.log (inOrder3 (Js.Undefined.return test1))
